### DISCUSSION TITLE
Cross-referencing panels inside a figure created all together (e.g., with `gridExtra::grid.arrange`)

### DISCUSSION THREAD

**Discussion by dmkaplan2000** [quarto-dev/quarto-cli#9194] [Q&A]
### Description

I have multi-panel figures created in a quarto document using `gridExtra::grid.arrange` and I want to cross-reference the panels in the figure. For example, if I have the following code:

````quarto
:::{#fig-myfig}
```{r}
df = data.frame(id=rep(c("(a) Line","(b) Parabola"),each=5),
                x = rep(-2:2,2), y = c(-2:2,(-2:2)^2))
library(ggplot2)
ggplot(df,aes(x,y)) + geom_line() + facet_wrap(vars(id))
```

A figure with multiple panels.
:::
````

Then I want to include in the document a cross-reference of the sort "Figure 1a". To reference the entire figure, I would put `@fig-myfig` so I would want to do something like `@fig-myfiga`, but of course that doesn't work as it doesn't recognize the `a` as being separate from `myfig`. I tried using the notation suggested in the quarto documentation of `@fig-myfig-1`, but this only works if quarto assembles the figure panels with subcaptions. For many reasons it is much easier for me to let `gridExtra::grid.arrange` or `facet_wrap` do the assembling. The only partial solutions I have found is to do `@fig-myfig a` or `@fig-myfig(a)`, but neither of these is the desired notation for figure panels. I also tried tricking quarto into separating `myfig` from `a`, for example doing `@fig-myfig``a`, but these tricks failed or didn't produce what I wanted. 

Is there a better way to achieve what I want?

**Comment by mcanouil** [comment]
If you combine the plots, there are no subfigures, thus Quarto cannot cross-reference what does not exist.

See:

- https://quarto.org/docs/authoring/cross-references-divs.html#subreferences
- https://quarto.org/docs/reference/cells/cells-knitr.html#figures

````qmd
---
title: "Quarto Playground"
format: html
---

```{r}
#| label: fig-myfig
#| fig-cap: A figure with multiple panels.
#| fig-subcap: true
#| layout-ncol: 2
df <- data.frame(
  id = rep(c("(a) Line", "(b) Parabola"), each = 5),
  x = rep(-2:2, 2),
  y = c(-2:2, (-2:2)^2)
)
library(ggplot2)
for (iid in unique(df[["id"]])) {
  print(
    ggplot(subset(df, id == iid)) +
      aes(x = x, y = y) +
      geom_line()
  )
}
```
````

**Reply by dmkaplan2000** [reply]
And to be clear, your solution would work in many cases, but in my specific case, it would be impossible or impractical for various reasons (I am using rather complex panel placement, legends, etc.).

**Reply by mcanouil** [reply]
> it would be impossible or impractical for various reasons (I am using rather complex panel placement, legends, etc.)

Well, maybe but `layout` can achieve "rather complex panel placement".

````qmd
---
title: "Quarto Playground"
format: html
---

```{r}
#| label: fig-myfig
#| fig-cap: A figure with multiple panels.
#| fig-subcap: true
#| layout: [[1, 1], [1], [2, 1]]
df <- data.frame(
  id = rep(LETTERS[1:5], each = 5),
  x = rep(-2:2, 5),
  y = c(-2:2, (-2:2)^2, (-2:2)^3, (-2:2)^4, (-2:2)^5)
)
library(ggplot2)
for (iid in unique(df[["id"]])) {
  print(
    ggplot(subset(df, id == iid)) +
      aes(x = x, y = y) +
      geom_line()
  )
}
```
````

**Comment by dmkaplan2000** [comment]
I get it, but how can I create the text "Fig. 1a" if I cannot put the `a` next to `@fig-myfig` without screwing up the cross-reference? That is my question more than anything, is there some way to tell quarto where the cross-reference ends without changing the final visual output? For example, in `bash`, when there is some risk of confusion regarding the extent of a shell variable, you can surround it in `{}` (e.g., `${HOME}abc` is valid, but `$HOMEabc` is not). Is there an equivalent for quarto?

**Reply by mcanouil** [reply]
>  For example, in bash, when there is some risk of confusion regarding the extent of a shell variable, you can surround it in {} (e.g., ${HOME}abc is valid, but $HOMEabc is not). Is there an equivalent for quarto?

No and I am pretty confident, it is not going to happen this way.

The only syntax that would make sense would be the one used by citeproc for references:

See <https://quarto.org/docs/authoring/footnotes-and-citations.html>.

(It is not supported, cc @cscheid )

````qmd
---
title: "Quarto Playground"
format: html
---

::: {#fig-myfig}
Figure

Content
:::

[@fig-myfig, a]
````

**Reply by dmkaplan2000** [reply]
This syntax seems interesting, but that notation for bibtex citations is really doing a different task I think, making it not particularly suitable in this use case in my view (but perhaps I am mistaken). In general, it would seem useful to be able to place something other than a space or a punctuation mark besides a cross-reference (as an aside, I tried `@fig-myfig.a` and this also failed). It is a really neat feature of quarto that you can often do paneling using quarto, but I think there will always be cases where quarto paneling isn't quite what you want, and ideally quarto should not be forcing users not to use very effective approaches to paneling that already exist in R.

**Reply by mcanouil** [reply]
The comma is not a separator.
Look at the link I shared for articles citation and the syntax.

> as an aside, I tried @fig-myfig.a and this also failed)

You can try a lot of other stuff, none of them will work. As I said, Quarto can only cross-reference "float objects" and only article citation supports (via Pandoc) some customisation of how you cite them.

**Reply by dmkaplan2000** [reply]
Perhaps there is some subtlety to how citeproc citations work that I am not aware of, but doesn't the comma become part of the output text, which is not what I am looking for (i.e., `[@fig-myfig, a]` would become "Figure 1, a", not "Figure 1a")? In any case, if this syntax could be made to work, then I am all for it. I wasn't proposing a specific solution based on what is done in bash or citeproc or anything else.

And in response to your next comment, I wasn't trying to get quarto to handle my cross-reference to a subpanel, I was just trying get the "a" adjacent to "Figure 1" in a somewhat natural way (e.g., without a strange space). I thought "." might work as a separator, but it didn't. I have moved on...

**Reply by mcanouil** [reply]
`[@fig-myfig, a]` this was only a rough example.
The syntax described in the link I shared is:

<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/8aae28dc-6fff-48c4-92a6-c8d756b551f6">

Anyhow, what you are asking is not currently possible and I am afraid it won't be soon.
I believe, the following might be of interest to you:
- https://github.com/quarto-dev/quarto-cli/discussions/8034

**Comment by gauravsk** [comment]
I found this while pondering the same question. I suspect there's a more "canonical" solution out there, but for now syntax `@fig-myfig\ignorespaces A` seems to do the trick for me.

**Reply by dmkaplan2000** [reply]
This is a good solution if you are rendering to PDF, but I do not believe it will work if rendering to other formats. In my case, I was rendering to DOCX. A solution that kinda works for DOCX and may work for other formats is to use an empty link to conceptually separate the "a" from the figure reference: `@fig-myfig[a]()` . This works OK, but it has formatting consequences for `a`. I tried using the notation to apply a `custom-style` to `a` to see if that would work, but this has the bizarre (erroneous?) consequence of just making the text disappear if it comes after a reference to a figure, even if separated from that reference by a space. For example:

```quarto
[a]{}
````

In a quarto document produces just the text "a" in the output DOCX, but:

```quarto
@fig-myfig[a]{}
```

**and**

```quarto
@fig-myfig [a]{} (NOTE THE SPACE)
```

Produce just "Figure 1" in the output DOCX file...
